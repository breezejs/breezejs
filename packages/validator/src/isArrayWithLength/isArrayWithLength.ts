/**
 * @author Mo Gusbi <me@mogusbi.co.uk>
 */
// tslint:disable no-any
import {registerDecorator, ValidationOptions} from 'class-validator';

export function IsArrayWithLength (min: number = 1, validationOptions?: ValidationOptions): Function {
  return (object: object, propertyName: string): void => {
    registerDecorator({
      constraints: [
        min
      ],
      name: 'IsArrayWithLength',
      options: validationOptions,
      propertyName: propertyName,
      target: object.constructor,
      validator: {
        validate (value: any): boolean {
          return value instanceof Array && value.length >= min;
        }
      }
    });
  };
}
