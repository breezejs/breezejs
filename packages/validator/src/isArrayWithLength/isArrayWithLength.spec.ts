/**
 * @author Mo Gusbi <me@mogusbi.co.uk>
 */
// tslint:disable no-any
import {Validator} from 'class-validator';
import {IsArrayWithLength} from './isArrayWithLength';

class TestDto {
  @IsArrayWithLength()
  public defaultArray: string[];

  @IsArrayWithLength(3)
  public customArray: string[];
}

describe('IsArrayWithLength', (): void => {
  let validator: Validator;

  beforeEach((): void => {
    validator = new Validator();
  });

  it('should fail if min length constraint is not met', async (): Promise<void> => {
    const dto: TestDto = Object.assign(new TestDto(), {
      customArray: [
        'test-1',
        'test-2'
      ],
      defaultArray: []
    });

    await expect(validator.validate(dto)).resolves.toMatchSnapshot();
  });

  it('should fail if value is not an array', async (): Promise<void> => {
    const dto: TestDto = Object.assign(new TestDto(), {
      customArray: <any>'test-1',
      defaultArray: [
        'test-1'
      ]
    });

    await expect(validator.validate(dto)).resolves.toMatchSnapshot();
  });

  it('should pass if min length constraint is met', async (): Promise<void> => {
    const dto: TestDto = Object.assign(new TestDto(), {
      customArray: [
        'test-1',
        'test-2',
        'test-3'
      ],
      defaultArray: [
        'test-1'
      ]
    });

    await expect(validator.validate(dto)).resolves.toMatchSnapshot();
  });
});
