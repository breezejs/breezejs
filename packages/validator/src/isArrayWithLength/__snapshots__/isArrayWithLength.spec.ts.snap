// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsArrayWithLength should fail if min length constraint is not met 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsArrayWithLength": "",
    },
    "property": "defaultArray",
    "target": TestDto {
      "customArray": Array [
        "test-1",
        "test-2",
      ],
      "defaultArray": Array [],
    },
    "value": Array [],
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsArrayWithLength": "",
    },
    "property": "customArray",
    "target": TestDto {
      "customArray": Array [
        "test-1",
        "test-2",
      ],
      "defaultArray": Array [],
    },
    "value": Array [
      "test-1",
      "test-2",
    ],
  },
]
`;

exports[`IsArrayWithLength should fail if value is not an array 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsArrayWithLength": "",
    },
    "property": "customArray",
    "target": TestDto {
      "customArray": "test-1",
      "defaultArray": Array [
        "test-1",
      ],
    },
    "value": "test-1",
  },
]
`;

exports[`IsArrayWithLength should pass if min length constraint is met 1`] = `Array []`;
